relay-wstring/src/lib.rs

```diff
@@ -0,0 +1,509 @@
+        let world = b"w\x00o\x00r\x00l\x00d\x00";
+        buf.copy_from_slice(world);
+        assert_eq!(b.as_slice(), world);
+    }
+
+    #[test]
+    fn test_wstr_get() {
+        // This is implemented with get_unchecked() so this is also already tested.
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+
+        let t = s.get(0..8).expect("expected Some(&WStr)");
+        assert_eq!(t.as_bytes(), b"h\x00e\x00l\x00l\x00");
+
+        let t = s.get(1..8);
+        assert!(t.is_none());
+    }
+
+    #[test]
+    fn test_wstr_get_mut() {
+        // This is implemented with get_unchecked_mut() so this is also already tested.
+        let mut b = Vec::from(&b"h\x00e\x00l\x00l\x00o\x00"[..]);
+        let s = unsafe { WStr::from_utf16le_unchecked_mut(b.as_mut_slice()) };
+
+        let t = s.get_mut(0..2).expect("expected Some(&mut Wstr)");
+        let buf = t.as_bytes_mut();
+        buf.copy_from_slice(b"x\x00");
+
+        assert_eq!(s.as_bytes(), b"x\x00e\x00l\x00l\x00o\x00");
+    }
+
+    #[test]
+    fn test_wstr_slice() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let sub = &s[2..8];
+        assert_eq!(sub.as_bytes(), b"e\x00l\x00l\x00");
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_wstr_bad_index() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let _r = &s[2..7];
+    }
+
+    #[test]
+    fn test_wstr_chars() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let chars: Vec<char> = s.chars().collect();
+        assert_eq!(chars, vec!['h', 'e', 'l', 'l', 'o']);
+
+        let b = b"\x00\xd8\x00\xdcx\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let chars: Vec<char> = s.chars().collect();
+        assert_eq!(chars, vec!['\u{10000}', 'x']);
+    }
+
+    #[test]
+    fn test_wstr_char_indices() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let chars: Vec<(usize, char)> = s.char_indices().collect();
+        assert_eq!(
+            chars,
+            vec![(0, 'h'), (2, 'e'), (4, 'l'), (6, 'l'), (8, 'o')]
+        );
+
+        let b = b"\x00\xd8\x00\xdcx\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let chars: Vec<(usize, char)> = s.char_indices().collect();
+        assert_eq!(chars, vec![(0, '\u{10000}'), (4, 'x')]);
+    }
+
+    #[test]
+    fn test_wstr_to_utf8() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
+        let out: String = s.to_utf8();
+        assert_eq!(out, "hello");
+    }
+
+    #[test]
+    fn test_wstr_is_ascii() {
+        let b = b"h\x00e\x00l\x00l\x00o\x00";
+        let s = unsafe { WStr::from_utf16le_unchecked(b) };
```

Comment:
Can't we use the safe version of this with unwrap instead, makes it easier to audit the code

Reply:
good call.  most these tests were written before the safe version existsed, hence why they looked like this.